exonum-java-binding/core/src/main/java/com/exonum/binding/core/blockchain/Blockchain.java

```diff
@@ -59,7 +157,46 @@
    */
   public static Blockchain newInstance(View view) {
     CoreSchema coreSchema = CoreSchema.newInstance(view);
-    return new Blockchain(coreSchema);
+    return new Blockchain(view, coreSchema);
+  }
+
+  /**
+   * Creates a proof for the block at the given height.
+   *
+   * <p>It allows creating genesis block proofs, but they make little sense, as a genesis
+   * block is supposed to be a "root of trust", hence, well-known to the clients verifying
+   * any subsequent proofs coming from the blockchain.
+   *
+   * <p>If you need to create a proof for a service index, use {@link #createIndexProof(String)}.
+   * @param blockHeight a height of the block for which to create a proof
+   * @throws IndexOutOfBoundsException if the height is not valid
+   * @see #createIndexProof(String)
+   */
+  public BlockProof createBlockProof(long blockHeight) {
+    checkHeight(blockHeight);
+    Proofs.BlockProof blockProof = BlockchainProofs.createBlockProof(view, blockHeight);
+    return BlockProof.newInstance(blockProof);
+  }
+
+  /**
+   * Creates a proof for a single index in the database.
+   * @param fullIndexName the full index name for which to create a proof
+   * @throws IllegalStateException if the view is not a snapshot, because a state of a service index
+   *     can be proved only for the latest committed block, not for any intermediate state during
+   *     transaction processing
+   */
+  public IndexProof createIndexProof(String fullIndexName) {
+    /*
+     todo: Shall we allow creating proofs for invalid (e.g., impossible) index names or throw
+      an exception?
+
+     todo: If index proofs for "uninitialized" indexes are forbidden, document that.
```

Comment:
The native will throw an exception, so I think we should document it.

Reply:
[But](https://github.com/exonum/exonum-java-binding/commit/fab544b8358bac6b1d201948e9f4782897cfd1e7#r36744088)

> Why doesn't it return a proof of absence (no such index)? Especially as an index is not "initialized" unless (?) something is recorded in it? Doesn't that place extra burden on the service developers to "initialize" an index? Shall they now write an extra "initialize" method in a schema of each service (and we — document that)?

Reply:
Actually, I don't know. I read sources of `proof_for_index` method - it looks like explicit check for existence of the index can be removed, and the method will return proof of absence. @slowli am I right?

Reply:
Yes, if the check of existence is removed, `proof_for_index` may return proof of absence for a non-existing index. The reason the method is implemented as is currently is that returning proof for a non-existing index doesn't make sense in a typical use case (returning the index proof + proof for data *within* the index, such as proof for a wallet in the cryptocurrency service).

Reply:
Thanks! Would you clarify please, if one first requests a proof from an empty index (= no wallets), will it become "initialized"? Will one be able to provide then a proof that the index remains empty?